loader/src/main/java/net/neoforged/fml/loading/ImmediateWindowHandler.java

```diff
@@ -32,7 +34,10 @@ public static void load(final String launchTarget, final String[] arguments) {
         } else {
             final var providername = FMLConfig.getConfigValue(FMLConfig.ConfigValue.EARLY_WINDOW_PROVIDER);
             LOGGER.info("Loading ImmediateWindowProvider {}", providername);
-            final var maybeProvider = ServiceLoader.load(ImmediateWindowProvider.class)
+            final var layer = Launcher.INSTANCE.findLayerManager()
+                    .flatMap(manager -> manager.getLayer(Layer.SERVICE))
+                    .orElse(ImmediateWindowHandler.class.getModule().getLayer()); // This will likely be the BOOT layer, but is a sensible fallback.
```

Comment:
Is it a sensible fallback? I would be in favor of throwing an exception early on to detect misconfiguration / unexpected environments.

Reply:
I believe so - the original overload of `ServiceLoader.load` uses the current thread's context class loader, which should be the same one as set by the BOOT layer in `BootstrapLauncher`: <https://github.com/McModLauncher/bootstraplauncher/blob/main/src/main/java/cpw/mods/bootstraplauncher/BootstrapLauncher.java#L172>

Reply:
I mean: is there any case in normal FML usage where the service layer would not be built by the team we reach that call? If not, we should throw. (Defensive programming and stuff)

Reply:
The call stack for a regular startup is something like this:
- BootstrapLauncher.main
- BootstrapLauncher.run
- BootstrapLauncherConsumer.accept
- Launcher.main
- Launcher.run
- TransformationServicesHandler.discoverServices
- ModDirTransformerDiscoverer.earlyInitialization
- ImmediateWindowHandler.load
And the SERVICE layer is built in `TransformationServicesHandler.discoverServices` just before calling `earlyInitialization`, which should make it impossible for it to happen. I personally don't expect `ImmediateWindowHandler.class.getModule().getLayer()` to ever be returned, but having it as a fallback should preserve the *current* behaviour if the order in which layers are built ever gets changed.

Reply:
Once we make the change that `ImmediateWindowProvider`s are loaded from the SERVICE layer, this is a feature that we must preserve. As such, crashing if the SERVICE layer is not available is the most appropriate thing to do, so that we can immediately find a bug should the code be moved to a different place.